### ISSUE TITLE
Invoke scheduler automatically when launching flux from flux

### ISSUE THREAD

**Issue by surajpkn** [flux-framework/flux-sched#60]
Until now, scheduler service is invoked through the 2 step commands:
`~/flux-core/src/cmd/flux -M ~/flux-sched/sched -C ~/flux-sched/rdl/\?.so -L ~/flux-sched/rdl/\?.lua start -s 3`
`~/flux-core/src/cmd/flux -M ~/flux-sched/sched -C ~/flux-sched/rdl/\?.so -L ~/flux-sched/rdl/\?.lua module load sched rdl-conf=../conf/hype.lua`

When flux launches flux, this needs to be automated. The most simplest way seems to be adding `sched[0]` to the `default_modules` in the broker. That way, the root instance as well as any child instance can directly start the sched module during startup. 
root:
`flux -M ~/flux-sched/ start -s2` 
child:
`flux -M ~/flux-sched/ broker`

I am not sure if thats the way to go for production, but it serves my immediate purposes. 

But a follow-up issue on that is the RDL. The scheduler needs to know its rdl-conf during startup. This can be achieved by having the root flux look into a default directory/file (`~/flux-sched/conf/default-conf.lua`) and the provide an additional argument to broker (`-C dir/conf.lua`) when the parent flux launches the child flux. That means that the parent scheduler prepares a conf for the child and then passes it as an argument. 

Another way to do that without having to extend the broker is: when the child scheduler starts up, it has to handshake with the parent (which is inevitable anyway for the dynamic scheduling). And as a part of the handshake, the parent can pass on the config (as json objects) to the child scheduler. 

I am more inclined to do the latter.

If there is a really correct way to do this, I would switch to that approach. Or if the right approach needs to evolve with time/other aspects, I would go ahead with this for now so that I can start with the scheduling problem. Please let me know your thoughts.

**Comment by garlick** [comment]
It may be easier to load and initialize the sched module from the initial program of an instance.  Things are a bit primitive right now, so a shell script that does that then starts work?  Practically speaking we don't want to hardwire a dependency on flux-sched in flux-core.

On becoming less primitive: one plan was to provide a tool to run as the initial program which has characteristics of sysVinit or systemd.  In other words have "modules" that can have dependencies, start in parallel, and ultimately enter some sort of productive mode after everything's ready, and finally shutdown.

Regarding the passing of RDL between instances, we have a notion of a per-program KVS space.  Could a chunk of RDL be placed there in the enclosing instance and then programs retrieve it via `parent-uri`?

In general, I think it's advisable if the parent never has to do anything in lock-step with a program it's running.  When you said "handshake" I worried about the coupling.  The parent must expect the child to misbehave - from the parent perspective it is "user code".

Anyway these are just thoughts - all of this is open for debate.

**Comment by surajpkn** [comment]
@garlick shell script to start the scheduler and RDL in parent-uri is possible. The handshake is required for the parent to store routing information about the child so that it can later contact the child by itself (e.g., to offer some resources).

**Comment by grondo** [comment]
> Regarding the passing of RDL between instances, we have a notion of a per-program KVS space. Could a chunk of RDL be placed there in the enclosing instance and then programs retrieve it via parent-uri?

This is almost necessary, at least in current design (see comment in #58). The parent has to place a subset of resources into kvs of all child programs describing the "container" anyway. Perhaps an issue should be open to have the ability for sched to read this as config in a child instance?

**Comment by grondo** [comment]
I think I see what you're saying, however the wreck replacement "shell" will have to read the same data to possibly do containment etc so as long as we don't make things any more difficult for that case.
It does help the simple bootstrap case if sched always reads config from the same spot, but I kind of feel like real-world configs may be too heavy to put the entire thing into kvs (maybe not though), which is kind of why I like multiple "config reader" methods for sched.

**Comment by grondo** [comment]
Oh, what if the config had *either* a file or a string of serialized rdl, sched loader could determine this at runtime? In general I do like that config for an instance including all modules could go in a common place, so I'd definitely be interested in exploring this idea. (I'd like to have @garlick's input as well)

**Comment by trws** [comment]
Either a file, a reference to a kvs location, or a string of RDL would be three that make sense to me. Maybe a flag for which loader to use also? I agree that there should be indirection for that too, it will save us headaches both now and later.



Sent with Good (www.good.com)

________________________________

**Comment by surajpkn** [comment]
Thanks for your comments! I also spoke to @SteVwonder and what I missed was that the scheduler seems to serialize the rdl and store it in the KVS. All that the child flux instance will have to do therefore, is lookup the kvs for the rdl using the parent handle and get that into its context. So no file needed! 
@lipari: @SteVwonder mentioned that you are working on a deserializer of the rdl? is this available already?

**Comment by dongahn** [comment]
This all makes sense to me. FWIW, the serialized rdl is placed by sched via jsc into `lwj.jobid.rdl`. You can update/fetch this field using [JSC_RDL macro](https://github.com/flux-framework/flux-core/tree/master/src/modules/libjsc).

@garlick, @surajpkn: I think the main reason that Suraj proposed a handshake is so that the parent sched can get the routing information by storing the message head to the job structure.

**Comment by dongahn** [comment]
@surajpkn several enhancements have been merged in related to this. Where does this issue stand?

**Comment by surajpkn** [comment]
I have taken the initial program approach to handle this and works fine. I think this issue can be closed.
